segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperServiceRunner.java

```diff
@@ -253,10 +258,37 @@ private void initializeZookeeper() throws Exception {
     private void runBookies() throws Exception {
         log.info("Starting Bookie(s) ...");
         for (int bkPort : this.bookiePorts) {
-            this.servers.add(runBookie(bkPort));
+            this.servers.add(runBookieWithRetry(bkPort));
         }
     }
 
+    private BookieServerAndResources runBookieWithRetry(int bkPort) throws Exception {
+        int count = 0;
+        boolean retry = true;
+        BookieServerAndResources bookieServerAndResources = null;
+        while (count < retries && retry) {
+            try {
+                log.info("Starting bookie with retry value of :" + count);
+                bookieServerAndResources = runBookie(bkPort);
+                retry = false;
+            } catch (BindException e) {
+                if ( ++count >= retries ) {
+                    e.printStackTrace();
+                    throw new BookieException(1) {
+                        @Override
+                        public String getMessage(int code) {
+                            return super.getMessage(code);
+                        }
+                    };
+                } else {
+                    bkPort = new Random().nextInt(MAX_PORT_COUNT);
```

Comment:
there is already a method to select the next available port efficiently in TestUtils#getAvailableListenPort. Can we check if we can reuse that?

Reply:
yes, checking on it.

Reply:
Right, please reuse the same logic so it is consistent across the board.

Reply:
done